### ISSUE TITLE
Handling remote connection close

### ISSUE THREAD

**Issue by gmr** [epgsql/epgsql#67]
I'm trying to figure out how to deal with a remote close of an epgsql or epgsqla connection async connection. 

In either event, instead of receiving a ``{epgsql, Connection, {notice, closed}}``, the backend is crashing:

```erlang
=ERROR REPORT==== 16-Jul-2015::22:18:10 ===
** Generic server <0.323.0> terminating 
** Last message in was {tcp,#Port<0.10271>,
                            <<69,0,0,0,109,83,70,65,84,65,76,0,67,53,55,80,48,
                              49,0,77,116,101,114,109,105,110,97,116,105,110,
                              103,32,99,111,110,110,101,99,116,105,111,110,32,
                              100,117,101,32,116,111,32,97,100,109,105,110,
                              105,115,116,114,97,116,111,114,32,99,111,109,
                              109,97,110,100,0,70,112,111,115,116,103,114,101,
                              115,46,99,0,76,50,56,55,50,0,82,80,114,111,99,
                              101,115,115,73,110,116,101,114,114,117,112,116,
                              115,0,0>>}
** When Server state == {state,gen_tcp,#Port<0.10271>,<<>>,
                               {64753,1424043724},
                               on_message,
                               {codec,[],[]},
                               {[],[]},
                               <0.224.0>,
                               [{<<"application_name">>,<<>>},
                                {<<"client_encoding">>,<<"UTF8">>},
                                {<<"DateStyle">>,<<"ISO, MDY">>},
                                {<<"integer_datetimes">>,<<"on">>},
                                {<<"IntervalStyle">>,<<"postgres">>},
                                {<<"is_superuser">>,<<"on">>},
                                {<<"server_encoding">>,<<"UTF8">>},
                                {<<"server_version">>,<<"9.4.4">>},
                                {<<"session_authorization">>,<<"gmr">>},
                                {<<"standard_conforming_strings">>,<<"on">>},
                                {<<"TimeZone">>,<<"US/Eastern">>}],
                               [],[],[],[],[],undefined,73}
** Reason for termination == 
** {empty,[{queue,get,[{[],[]}],[{file,"queue.erl"},{line,182}]},
           {epgsql_sock,command_tag,1,
                        [{file,"../epgsql-wrapper/epgsql-git/src/epgsql_sock.erl"},
                         {line,409}]},
           {epgsql_sock,on_message,2,
                        [{file,"../epgsql-wrapper/epgsql-git/src/epgsql_sock.erl"},
                         {line,684}]},
           {epgsql_sock,loop,1,
                        [{file,"../epgsql-wrapper/epgsql-git/src/epgsql_sock.erl"},
                         {line,334}]},
           {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,593}]},
           {gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,659}]},
           {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]}
```

I'm using LISTEN/NOTIFY with async registered to a gen_server and trying to handle error conditions.

Are my expectations off or am I doing something wrong?

**Comment by davidw** [comment]
This isn't something I've ever used myself.

**Comment by gmr** [comment]
The easiest way to see the behavior is to have an app connect to Postgres and then in psql run a ``select  pg_cancel_backend(PID)`` or ``select pg_terminated_backend(PID)``.

My expectation based upon the README is that `` {'EXIT', C, _Reason}`` would cause the termination:

```erlang
  {ok, C} = epgsqla:start_link(),
  Ref = epgsqla:connect(C, "localhost", "username", "psss", [{database, "test_db"}]),
  receive
    {C, Ref, connected} ->
        {ok, C};
    {C, Ref, Error = {error, _}} ->
        Error;
    {'EXIT', C, _Reason} ->
        {error, closed}
  end.
```

Not:
```erlang
** Reason for termination == 
** {empty,[{queue,get,[{[],[]}],[{file,"queue.erl"},{line,182}]},
           {epgsql_sock,command_tag,1,
                        [{file,"../epgsql-wrapper/epgsql-git/src/epgsql_sock.erl"},
                         {line,409}]},
           {epgsql_sock,on_message,2,
                        [{file,"../epgsql-wrapper/epgsql-git/src/epgsql_sock.erl"},
                         {line,684}]},
           {epgsql_sock,loop,1,
                        [{file,"../epgsql-wrapper/epgsql-git/src/epgsql_sock.erl"},
                         {line,334}]},
           {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,593}]},
           {gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,659}]},
           {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]}
```

**Comment by ostinelli** [comment]
I wonder if this is also related to what I'm experiencing in https://github.com/epgsql/epgsql/issues/79.

**Comment by seriyps** [comment]
Was fixed in #89
